Tools/Scripts/libraries/webkitcorepy/webkitcorepy/credentials.py

```diff
@@ -51,36 +51,49 @@ def credentials(url, required=True, name=None, prompt=None, key_name='password')
 
     username_prompted = False
     key_prompted = False
-    if not username:
-        try:
-            if keyring:
-                username = keyring.get_password(url, 'username')
-        except (RuntimeError, AttributeError):
-            pass
 
-        if not username and required:
-            if not sys.stderr.isatty() or not sys.stdin.isatty():
-                raise OSError('No tty to prompt user for username')
-            sys.stderr.write("Authentication required to use {}\n".format(prompt or name))
-            sys.stderr.write('Username: ')
-            username = Terminal.input()
-            username_prompted = True
+    for attempt in range(retry):
+        if attempt:
+            sys.stderr.write('Ignoring keychain values and re-prompting user\n')
+        if not username:
+            try:
+                if keyring and not attempt:
+                    username = keyring.get_password(url, 'username')
+            except (RuntimeError, AttributeError):
+                pass
 
-    if not key:
-        try:
-            if keyring:
-                key = keyring.get_password(url, username)
-        except (RuntimeError, AttributeError):
-            pass
+            if not username and required:
+                if not sys.stderr.isatty() or not sys.stdin.isatty():
+                    raise OSError('No tty to prompt user for username')
+                sys.stderr.write("Authentication required to use {}\n".format(prompt or name))
+                sys.stderr.write('Username: ')
+                username = Terminal.input()
+                username_prompted = True
 
-        if not key and required:
-            if not sys.stderr.isatty() or not sys.stdin.isatty():
-                raise OSError('No tty to prompt user for username')
-            key = getpass.getpass('{}: '.format(key_name.capitalize()))
-            key_prompted = True
+        if not key:
+            try:
+                if keyring and not attempt:
+                    key = keyring.get_password(url, username)
+            except (RuntimeError, AttributeError):
+                pass
 
-    if username and key:
-        _cache[name] = (username, key)
+            if not key and required:
+                if not sys.stderr.isatty() or not sys.stdin.isatty():
+                    raise OSError('No tty to prompt user for username')
+                key = getpass.getpass('{}: '.format(key_name.capitalize()))
+                key_prompted = True
+
+        if username and key and (not validater or validater(username, key)):
+            _cache[name] = (username, key)
+            break
+
+        username = None
+        key = None
+        sys.stderr.write("Failed to validate credentials for '{}' ({} attempt)\n".format(url, string_utils.ordinal(attempt + 1)))
+
+    if not username or not key:
+        sys.stderr.write("Exhausted attempts to prompt user for '{}' credentials\n".format(url))
+        sys.exit(1)
```

Comment:
I think it's the right call here because we're already hijacking the terminal. Usually I'd agree, but I can't imagine programs calling this one regularly, and we have some precedent for immediately exiting in user facing code like this, even in a module, `webkitcorepy/terminal.py`'s `input` function comes to mind.